38_EXRSplit/main.cpp

```diff
@@ -1,94 +1,89 @@
-	else
-	{
-		logger->log("To many arguments - pass a single filename without .exr extension of OpenEXR image placed in media/OpenEXR!", ILogger::ELL_ERROR);
-		return 0;
-	}
+		// Remember to call the base class initialization!
+		if (!base_t::onAppInitialized(std::move(system)))
+			return false;
 
-	IAssetLoader::SAssetLoadParams loadParams;
-	constexpr std::string_view defaultImagePath = "../../media/noises/spp_benchmark_4k_512.exr";
-	const auto filePath = std::string(isItDefaultImage ? defaultImagePath.data() : argv[1]);
 
-	const asset::COpenEXRMetadata* meta;
-	auto image_bundle = assetManager->getAsset(filePath, loadParams);
-	auto contents = image_bundle.getContents();
-	{
-		bool status = !contents.empty();
-		assert(status);
-		status = meta = image_bundle.getMetadata()->selfCast<const COpenEXRMetadata>();
-		assert(status);
-	}
+		// Nabla's virtual filesystem has no notion of a Current Working Directory as its inherently thread-unsafe
+		// everything operates on "absolute" paths
+		const nbl::system::path CWD = path(argv[0]).parent_path().generic_string() + "/";
 
-	uint32_t i = 0u;
-	for (auto asset : contents)
-	{
-		auto image = IAsset::castDown<ICPUImage>(asset);
-		const auto* metadata = static_cast<const COpenEXRMetadata::CImage*>(meta->getAssetSpecificMetadata(image.get()));
+		// we assume you'll run the example `../..` relative to our media dir
+		path mediaWD = CWD.generic_string() + "../../media/";
 
-		ICPUImageView::SCreationParams imgViewParams;
-		imgViewParams.flags = static_cast<ICPUImageView::E_CREATE_FLAGS>(0u);
-		imgViewParams.image = std::move(image);
-		imgViewParams.format = imgViewParams.image->getCreationParameters().format;
-		imgViewParams.viewType = ICPUImageView::ET_2D;
-		imgViewParams.subresourceRange = { static_cast<IImage::E_ASPECT_FLAGS>(0u),0u,1u,0u,1u };
-		auto imageView = ICPUImageView::create(std::move(imgViewParams));
+		auto assetManager = make_smart_refctd_ptr<nbl::asset::IAssetManager>(smart_refctd_ptr(m_system));
 
-		auto channelsName = metadata->m_name;
-
-		std::filesystem::path filename, extension;
-		core::splitFilename(filePath.c_str(), nullptr, &filename, &extension);
-		const std::string finalFileNameWithExtension = filename.string() + extension.string();
-		const std::string finalOutputPath = channelsName.empty() ? (filename.string() + "_" + std::to_string(i++) + extension.string()) : (filename.string() + "_" + channelsName + extension.string());
+		// when assets are retrieved you need to provide parameters that control the loading process
+		nbl::asset::IAssetLoader::SAssetLoadParams lp;
+		// at the very least you need to provide the `workingDirectory` if your asset depends on others
+		// this helps resolve relative paths for things such as textures
+		lp.workingDirectory = mediaWD;
+		
 
-		const auto writeParams = IAssetWriter::SAssetWriteParams(imageView.get(), EWF_BINARY);
+		auto checkedLoad = [&]<class T>(const string & key)->smart_refctd_ptr<T>
```

Comment:
maybe undo the lambda and remove the templating, you're looking for only one asset type (ICPUImage)